### DISCUSSION TITLE
Variability between property name prefixing

### DISCUSSION THREAD

**Discussion by lehn0058** [SPSCommerce/sps-api-standards#6] [Q&A] [answered]
I'm wanting to discuss property naming mentioned in the following section: https://atlassian.spscommerce.com/wiki/display/SIP/API+-+Naming#APINaming-PropertyNames

It mentions rules for when to prefix property names vs when not too. Below is an example:

```
// INCORRECT EXAMPLE
GET /v1/books/2345
RESPONSE
{
    "bookId": 2345,                             // since the request for the resource was a "book", the attribute should just be "id"
    "name": "The Best Book",
    "author": {                    
        "authorId":456,                         // the parent entity is an author, so it should be "id" for the attribute name.
        "authorName": "John Doe",               // the parent entity is an author, so it should be "name" for the attribute name.
        "publisherName": "My Publishing House"  // this name attribute MUST have a prefix for another entity, since its not the name of this entity "author"
        "agentId": 2                            // this name is appropriate with Id following its associated entity name.
    }
}
 
// CORRECT EXAMPLE
GET /v1/books/2345
RESPONSE
{
    "id": 2345,                            
    "name": "The Best Book",
    "author": {                    
        "id":456,                      
        "name": "John Doe",            
        "publisherName": "My Publishing House",
        "agentId": 2
    }
}
```

While this does make the API response look cleaner, I've often experienced that this causes more confusion for the users of the API, in particular for the unique identifier (id in this case). For example, if we were comparing two objects with a foreign key pointing to each other:

```
const item = {
    id: 2345,                            
    name: "my cool name",
}

const thing = {
   id: 7890,
   bookId: 2345,
   count: 20
}
```

In this example, I have no idea looking at the spec what object types item or thing is. If we are doing an equality and if we know that we have a foreign key reference between the two objects (no guarantee from the schema though), looking at the data this example maybe indicates that we should do:
```
item.id == thing.bookId
```

But when in our code editor we often are not looking at the JSON data. We are looking at the properties on the response object for the spec we are given or the object types we have. It is much more clear if the primary identifiers had been fully prefixed out and the properties of both objects match:

```
const item = {
    bookId: 2345,                            
    name: "my cool name",
}

const thing = {
   publisherId: 7890,
   bookId: 2345,
   count: 20
}
```

Even though `item` may still not be a book (`thing` has multiple ids after all) there is still much less confusion about whether we are doing the correct equality comparison:
```
item.bookId == thing.bookId
```

**Comment by travisgosselin** [comment, marked as answer]
The goal of the existing standards was to provide consistent naming on properties and what they are associated to for not just IDs but for all property types. Are you advocating that "id" is an exception to that rule only? or that ALL properties should have prefixes?

I think I prefer that consistency on all properties without the prefix on any, than having an exception for id. I cannot say I have experienced the confusion you have in other APIs that have taken this approach. Even in your example when using `item.id` that is obvious to me that it is the `id` of the object (item). If the variable was poorly named... obviously thats a different issue. 

Looking at some of our reference APIs, i.e. GitHub API as a good place to start, as its developer experience rep is usually pretty high, their endpoints follow the existing pattern specified in the standards (i.e. using ID). 
https://docs.github.com/en/rest/reference/

**Reply by eggilbert** [reply]
The other thing to consider is that the internal model that is used to consume the data can be decoupled from the the JSON property names used in (de-)serialization. So, in the example above the field could be called `bookId` in the code but map to `id` in the input/output where it will be used in the context of a larger response object. 🤷‍♂️

**Reply by lehn0058** [reply]
@eggilbert That is very true, though this is easier done in the backend than on the frontend. JSON is a javascript/typescript object so as soon as you get it from your API call it is already parsed for you as an object. There is not an easy way to intercept this/change what properties map to (no simple use of serialization annotations). Unless you do a separate transformation/map step, the values that the API sends is usually what the UI gets to deal with.

**Reply by lehn0058** [reply]
@travisgosselin I'm not necessarily saying that id should be an exception, nor that we should prefix all attributes, more wanting to point out a piece of confusion from other API's I have worked with specifically around sometimes just calling object unique identifiers `id`. While I would love it if variable names someone else (or myself 12 months ago) always made perfect sense, I don't think we should assume they always will. Having the API be more clear as to what the identifiers are helps users who come into code for the first time have a more clear definition of what identifiers are the same. I agree that consistency is super important, but I would argue that sometimes having an item id be `id` on one object and `itemId` on another is not actually consistency.

**Reply by travisgosselin** [reply]
Thinking of this modification, its definitely a change, but I'm not clear on if it is a change that has concrete benefits beyond preference. In those circumstances, I tend to lean on other respectable examples (like GitHub, etc). For comparison purposes, do you have some other API examples from "good external APIs" that would draw on this consistency? Without tangible benefits, or respected examples I'm hesitant to say we should continue with this much further? Thoughts...?

**Reply by lehn0058** [reply]
It may be worth getting input from those in the UI guild - curious if others who do mostly UI have similar (or completely opposite) opinions.

**Reply by turnerjw** [reply]
I'm more concerned about lazily prefixed properties (ex. `itemId`, `itemBriefId`, etc.), where differently named properties are actually the same thing but are named that way just to follow a pattern (incorrectly).  I think that introduces a bit of doubt where I would question whether something is mistakenly or deliberately named.  

As far as transforming data on the front end, it shouldn't be too difficult.  With API standards we should be able to quickly form patterns for transformations.  Also, any team using react-query can take advantage of it's `select` api (https://react-query.tanstack.com/reference/useQuery).  React-query is great 😄

**Reply by travisgosselin** [reply]
When I think of Item vs ItemBrief naming conventions, I guess those seem more like "model" names to me than resource names in the path. But nevertheless, I think your point is appreciated anyhow.

**Reply by lehn0058** [reply]
React query (and any other future tech like it) is pretty cool. I was unaware that it let you map the data coming in to a different structure similar to how json annotations do on the backend, but if that is the case then my concern probably doesn't matter as much. It will be on each consumer of the API to use it how they want (or map the properties to what makes sense for them). Keeping the flexibility while also keeping the schema clean feels like a win-win.

**Reply by acchristensen** [reply]
Reading through this thread, I do go back and forth.  I personally prefer the explicit naming of the id, but I also appreciate that it's important to unequivocally say which identifier (if there are multiple for associations to other objects) is referring to the `this` or `self`.  The callout that rolling in some kind of prefix that indicates the model is a good one... and that actually does make second guess my preference.  The identifier shouldn't probably be leaned on for the data model.  It does kinda play in to my other concern of recognizing the API response body out-of-context (see the other discussion #5 ).  Another thing that might help alleviate this is if the identifier itself is "self-describing" :). Not being a troll, but just calling out a case where that could be helpful is what is brought up in this discussion.

**Reply by lehn0058** [reply]
Lol - I don't take that as trolling at all, lets call it cross referencing :-)

I do feel like a self describing identifier doesn't actually solve the particular issue I have run into. Reason being you only see it when you are looking at the API spec or at data coming in over the wire. You don't see the self-describing id's value when you are using the object model in your code files that has already been created by someone. It can be used to help identify an issue as you are debugging if a mismatch was made, but it does not help to avoid the initial potential bug.

You did mention an interesting point about using `id` though, as it does become the standard way to say "this is the unique identifier of the current object". You do loose that if we always prefix the ids.

**Reply by acchristensen** [reply]
Gr.  Yeah, it's a good point since we're focusing on the clarity of the spec.
One convention we do see sometimes is to specially annotate a json field with a `$` or `@` prefix.  So we could do something like `@spsItemId` to indicate that the ID of `this` is that particular attribute.  But... that feels a little esoteric :-/

**Reply by travisgosselin** [reply]
Agreed - I think adding "sps" as a prefix likely captures all "@" or symbol advantages and then some. I'd like to reserve symbol prefixes for potential usage in REST specific semantics (generically) in the future as well (not sure how, just thinking long term out loud).

**Reply by acchristensen** [reply]
Well, except the `sps` prefix could be for multiple identifiers in the body.  So, it could both be a prefix of an identifier of `this`, but then also be the prefix of identifiers of any referenced objects.

**Reply by travisgosselin** [reply]
@lehn0058 - just following up on the discussion - do you feel like this one has come to a sufficient answer? Leaving the standards as is? Or want to leave the discussion open for further analysis?

**Reply by lehn0058** [reply]
I feel like we have a workaround (let the clients override if they decide to), but I don't really feel like we have a solution to this issue. That said, I'm not sure this is an easy one to solve for... most specs probably go the route of "this is what most other ones are doing" which is why I have run into this issue with so many of them. If so many others have not been able to solve it though, I'm not confident we will be able to without significant investment, and that doesn't really feel worth it for something like this. I am ok with closing out this discussion.
